File: kernel/drivers/tty/keyboard.c
#include "drivers/tty/keyboard.h"

#include "drivers/tty/tty.h"
#include "drivers/tty/virtterm.h"

#include "main/io.h"
#include "main/interrupt.h"

#include "util/debug.h"

#define IRQ_KEYBOARD 1

/* Indicates that one of these is "being held down" */
#define SHIFT_MASK 0x1
#define CTRL_MASK  0x2
/* Indicates that an escape code was the previous key received */
#define ESC_MASK   0x4
static int curmask = 0;

/* Where to read from to get scancodes */
#define KEYBOARD_IN_PORT 0x60
#define KEYBOARD_CMD_PORT 0x61

/* Scancodes for special keys */
#define LSHIFT  0x2a
#define RSHIFT  0x36
#define CTRL    0x1d
/* Right ctrl is escaped */
/* Our keyboard driver totally ignores ALT */
#define ESC0    0xe0
#define ESC1    0xe1

/* Special stuff for scrolling (note that these only work when ctrl is held) */
#define SCROLL_UP   0x0e
#define SCROLL_DOWN 0x1c

/* Keys for switching virtual terminals - for now F1,F2,...,F10 */
#define VT_KEY_LOW  0x3b
#define VT_KEY_HIGH 0x44

/* If the scancode & BREAK_MASK, it's a break code; otherwise, it's a make code */
#define BREAK_MASK 0x80

#define NORMAL_KEY_HIGH 0x39

/* Some sneaky value to indicate we don't actually pass anything to the terminal */
#define NO_CHAR 0xff

/* Scancode tables copied from
   http://www.win.tue.nl/~aeb/linux/kbd/scancodes-1.html */

/* The scancode table for "normal" scancodes - from 02 to 39 */
/* Unsupported chars are symbolized by \0 */
static const char *normal_scancodes = "\0"              /* Error */
                                      "\e"              /* Escape key */
                                      "1234567890-="    /* Top row */
                                      "\b"              /* Backspace */
                                      "\tqwertyuiop[]\n" /* Next row - ish */
                                      "\0"              /* Left ctrl */
                                      "asdfghjkl;\'`"
                                      "\0"              /* Lshift */
                                      "\\"
                                      "zxcvbnm,./"
                                      "\0\0\0"          /* Rshift, prtscrn, Lalt */
                                      " ";              /* Space bar */
/* As above, but if shift is pressed */
static const char *shift_scancodes = "\0"
                                     "\e"
                                     "!@#$%^&*()_+"
                                     "\b"
                                     "\tQWERTYUIOP{}\n"
                                     "\0"
                                     "ASDFGHJKL:\"~"
                                     "\0"
                                     "|"
                                     "ZXCVBNM<>?"
                                     "\0\0\0"
                                     " ";

static keyboard_char_handler_t keyboard_handler = NULL;

/* This is the function we register with the interrupt handler - it reads the
 * scancode and, if appropriate, call's the tty's receive_char function */
static void
keyboard_intr_handler(regs_t *regs)
{
        uint8_t sc; /* The scancode we receive */
        int break_code; /* Was it a break code */
        /* the resulting character ('\0' -> ignored char) */
        uint8_t c = NO_CHAR;
        /* Get the scancode */
        sc = inb(KEYBOARD_IN_PORT);

        /* Separate out the break code */
        break_code = sc & BREAK_MASK;
        sc &= ~BREAK_MASK;

        /* dbg(DBG_KB, ("scancode 0x%x, break 0x%x\n", sc, break_code)); */

        /* The order of this conditional is very, very tricky - be careful when
         * editing! */

        /* Most break codes are ignored */
        if (break_code) {
                /* Shift/ctrl release */
                if (sc == LSHIFT || sc == RSHIFT)
                        curmask &= ~SHIFT_MASK;
                else if (sc == CTRL)
                        curmask &= ~CTRL_MASK;
        }
        /* Check for the special keys */
        else if (sc == LSHIFT || sc == RSHIFT) {
                curmask |= SHIFT_MASK;
        } else if (sc == CTRL) {
                curmask |= CTRL_MASK;
        }
        /* All escaped keys past this point (anything except right shift and right
         * ctrl) will be ignored */
        else if (curmask & ESC_MASK) {
                /* Escape mask only lasts for one key */
                curmask &= ~ESC_MASK;
        }
        /* Now check for escape code */
        else if (sc == ESC0 || sc == ESC1) {
                curmask |= ESC_MASK;
        }
        /* Check for Ctrl+Fn key which indicates virt term switch */
        else if (sc >= VT_KEY_LOW && sc <= VT_KEY_HIGH) {
                vt_switch(sc - VT_KEY_LOW);
        }
        /* Check for Ctrl+Backspace which indicates scroll down */
        else if ((curmask & CTRL_MASK) && sc == SCROLL_DOWN) {
                vt_scroll(DISPLAY_HEIGHT, 0);
        }
        /* Check for Ctrl+Enter which indicates scroll down */
        else if ((curmask & CTRL_MASK) && sc == SCROLL_UP) {
                vt_scroll(DISPLAY_HEIGHT, 1);
        }
        /* Check to make sure the key isn't high enough that it won't be found in
         * tables */
        else if (sc > NORMAL_KEY_HIGH) {
                /* ignore */
        }
        /* Control characters */
        else if (curmask & CTRL_MASK) {
                /* Because of the way ASCII works, the control chars are based on the
                 * values of the shifted chars produced without control */
                c = shift_scancodes[sc];
                /* Range of chars that have corresponding control chars */
                if (c >= 0x40 && c < 0x60)
                        c -= 0x40;
                else
                        c = NO_CHAR;
        }
        /* Capitals */
        else if (curmask & SHIFT_MASK) {
                c = shift_scancodes[sc];
        } else {
                c = normal_scancodes[sc];
        }

        /* Give the key to the vt system, which passes it to the tty */
        if (c != NO_CHAR && keyboard_handler) {
                keyboard_handler(c);
        }
}

void
keyboard_init()
{
        intr_map(IRQ_KEYBOARD, INTR_KEYBOARD);
        intr_register(INTR_KEYBOARD, keyboard_intr_handler);
}

void
keyboard_register_handler(keyboard_char_handler_t handler)
{
        keyboard_handler = handler;
}

File: kernel/drivers/bytedev/tty/keyboard.rs

//! The Reenix keyboard support.
#![allow(dead_code)]

use core::prelude::*;
use procs::interrupt;
use base::io::inb;

/* Indicates that one of these is "being held down" */
const SHIFT_MASK : u8 = 0x1;
const CTRL_MASK  : u8 = 0x2;
/* Indicates that an escape code was the previous key received */
const ESC_MASK   : u8 = 0x4;

/* Where to read from to get scancodes */
const KEYBOARD_IN_PORT : u16 = 0x60;
const KEYBOARD_CMD_PORT : u16 = 0x61;

/* Scancodes for special keys */
const LSHIFT  : u8 = 0x2a;
const RSHIFT  : u8 = 0x36;
const CTRL    : u8 = 0x1d;
/* Right ctrl is escaped */
/* Our keyboard driver totally ignores ALT */
const ESC0    : u8 = 0xe0;
const ESC1    : u8 = 0xe1;

/* Special stuff for scrolling (note that these only work when ctrl is held) */
const SCROLL_UP   : u8 = 0x0e;
const SCROLL_DOWN : u8 = 0x1c;

/* Keys for switching virtual terminals - for now F1,F2,...,F10 */
const VT_KEY_LOW  : u8 = 0x3b;
const VT_KEY_HIGH : u8 = 0x44;

/* If the scancode & BREAK_MASK, it's a break code; otherwise, it's a make code */
const BREAK_MASK : u8 = 0x80;

const NORMAL_KEY_HIGH : u8 = 0x39;

/* Some sneaky value to indicate we don't actually pass anything to the terminal */
const NO_CHAR : u8 = 0xff;

/* Scancode tables Based on
 *  http://www.win.tue.nl/~aeb/linux/kbd/scancodes-1.html
 */

/* The scancode table for "normal" scancodes - from 02 to 39 */
/* Unsupported chars are symbolized by \x00 */
const NORMAL_SCANCODES : &'static str = concat!("\x00",             /* Error */
                                                "\x1b",             /* Escape key */
                                                "1234567890-=",     /* Top row */
                                                "\x08",             /* Backspace */
                                                "\tqwertyuiop[]\n", /* Next row - ish */
                                                "\x00",             /* Left ctrl */
                                                "asdfghjkl;'`",
                                                "\x00",             /* Lshift */
                                                "\\",
                                                "zxcvbnm,./",
                                                "\x00\x00\x00",     /* Rshift, prtscrn, Lalt */
                                                " "                 /* Space bar */);

const SHIFT_SCANCODES  : &'static str = "\x00\x1b!@#$%^&*()_+\x08\tQWERTYUIOP{}\n\x00ASDFGHJKL:\"~\x00|ZXCVBNM<>?\x00\x00\x00 ";
// For some reason the parser does not like the much easier to read concat version.
                                      /*concat!("\x00",
                                                "\x1b",
                                                "!@#$%^&*()_+"
                                                "\x08",
                                                "\tQWERTYUIOP{}\n",
                                                "\x00",
                                                "ASDFGHJKL:\"~",
                                                "\x00",
                                                "\\",
                                                "ZXCVBNM<>?",
                                                "\x00\x00\x00",
                                                " ");*/

pub enum KeyboardEvent {
    Normal(u8), Switch(u8), ScrollUp, ScrollDown,
}

const IRQ_KEYBOARD : u16 = 0x1;

pub fn init_stage1() {}

pub fn init_stage2() {
    interrupt::map(IRQ_KEYBOARD, interrupt::KEYBOARD);
    interrupt::register(interrupt::KEYBOARD, keyboard_handler);
}

pub type KeyboardHandler = extern "Rust" fn(KeyboardEvent);

pub struct Keyboard { curmask: u8, handler: KeyboardHandler }

fn default_handler(_: KeyboardEvent) { }
static mut KEYBOARD : Keyboard = Keyboard { curmask : 0, handler: default_handler };

#[inline]
pub fn get_keyboard() -> &'static mut Keyboard { unsafe { &mut KEYBOARD } }

impl Keyboard {
    pub fn set_handler(&mut self, handler: KeyboardHandler) {
        self.handler = handler;
    }

    pub fn handle_interrupt(&mut self) {
        let inp = unsafe { inb(KEYBOARD_IN_PORT) };
        let sc = inp & !BREAK_MASK;
        let c = if inp & BREAK_MASK != 0 {
            /* Most break codes are ignored */
            /* Shift/ctrl release */
            if sc == LSHIFT || sc == RSHIFT {
                self.curmask &= !SHIFT_MASK;
            } else if sc == CTRL {
                self.curmask &= !CTRL_MASK;
            }
            return;
        } else if sc == LSHIFT || sc == RSHIFT {
            self.curmask |= SHIFT_MASK;
            return;
        } else if sc == CTRL {
            self.curmask |= CTRL_MASK;
            return;
        } else if self.curmask & ESC_MASK != 0 {
            /* Escape mask only lasts for one key */
            self.curmask &= !ESC_MASK;
            return;
        } else if sc == ESC0 || sc == ESC1 {
            self.curmask |= ESC_MASK;
            return;
        } else if sc >= VT_KEY_LOW && sc <= VT_KEY_HIGH {
            Switch(sc - VT_KEY_LOW)
        } else if (self.curmask & CTRL_MASK) != 0 && sc == SCROLL_DOWN {
            ScrollDown
        } else if (self.curmask & CTRL_MASK) != 0 && sc == SCROLL_UP {
            ScrollUp
        } else if sc > NORMAL_KEY_HIGH {
            return;
        } else if self.curmask & CTRL_MASK != 0 {
            /* Because of the way ASCII works, the control chars are based on the
             * values of the shifted chars produced without control */
            let c = SHIFT_SCANCODES.as_bytes()[sc as uint];
            /* Range of chars that have corresponding control chars */
            if c >= 0x40 && c < 0x60 {
                Normal(c - 0x40)
            } else {
                return;
            }
        } else if self.curmask & SHIFT_MASK != 0 {
            Normal(SHIFT_SCANCODES.as_bytes()[sc as uint])
        } else {
            Normal(NORMAL_SCANCODES.as_bytes()[sc as uint])
        };
        /* Give the key to the vt system, which passes it to the tty */
        let h = self.handler;
        h(c);
    }
}

extern "Rust" fn keyboard_handler(_: &mut interrupt::Registers) {
    get_keyboard().handle_interrupt();
}
